app/design-system/design-system-hedvig/src/main/kotlin/com/hedvig/android/design/system/hedvig/Button.kt

```diff
@@ -0,0 +1,382 @@
+
+@Composable
+fun HedvigButton(
+  text: String,
+  onClick: () -> Unit,
+  enabled: Boolean,
+  modifier: Modifier = Modifier,
+  buttonStyle: ButtonDefaults.ButtonStyle = ButtonDefaults.buttonStyle,
+  buttonSize: ButtonDefaults.ButtonSize = ButtonDefaults.buttonSize,
+  interactionSource: MutableInteractionSource? = null,
+  isLoading: Boolean = false,
+) {
+  HedvigButton(
+    onClick = onClick,
+    enabled = enabled,
+    modifier = modifier,
+    buttonStyle = buttonStyle,
+    buttonSize = buttonSize,
+    interactionSource = interactionSource,
+  ) {
+    val buttonColors = buttonStyle.style.buttonColors
+    val loadingTransition = updateTransition(isLoading, label = "loading transition")
+    loadingTransition.AnimatedContent(
+      transitionSpec = {
+        fadeIn(tween(durationMillis = 220, delayMillis = 90)) togetherWith fadeOut(tween(90))
+      },
+      contentAlignment = Alignment.Center,
+    ) { loading ->
+      if (loading) {
+        LayoutWithoutPlacement(
+          sizeAdjustingContent = { HedvigText(text = text, modifier = Modifier.withoutPlacement()) },
+        ) {
+          ThreeDotsLoading(
+            stableColor = buttonColors.activeLoadingIndicatorColor,
+            temporaryColor = buttonColors.inactiveLoadingIndicatorColor,
+            modifier = Modifier.align(Alignment.Center),
+          )
+        }
+      } else {
+        HedvigText(text = text)
+      }
+    }
+  }
+}
+
+@Composable
+fun HedvigButton(
+  onClick: () -> Unit,
+  enabled: Boolean,
+  modifier: Modifier = Modifier,
+  buttonStyle: ButtonDefaults.ButtonStyle = ButtonDefaults.buttonStyle,
+  buttonSize: ButtonDefaults.ButtonSize = ButtonDefaults.buttonSize,
+  interactionSource: MutableInteractionSource? = null,
+  content: @Composable RowScope.() -> Unit,
+) {
+  @Suppress("NAME_SHADOWING")
+  val interactionSource = interactionSource ?: remember { MutableInteractionSource() }
+  val buttonColors = buttonStyle.style.buttonColors
+  val containerColor = buttonColors.containerColor(enabled)
+  val contentColor = buttonColors.contentColor(enabled)
+  val isHovered by interactionSource.collectIsHoveredAsState()
+  val hoverColor = buttonColors.hoverColor(isHovered)
+  val color by animateColorAsState(
+    if (isHovered) {
+      hoverColor.compositeOver(containerColor)
+    } else {
+      containerColor
+    },
+  )
+  Surface(
+    onClick = onClick,
+    modifier = modifier.semantics { role = Role.Button },
+    enabled = enabled,
+    shape = buttonSize.size.shape,
+    color = color,
+    contentColor = contentColor,
+    interactionSource = interactionSource,
+  ) {
+    ProvideTextStyle(buttonSize.size.textStyle) {
+      Row(
+        modifier = Modifier.padding(buttonSize.size.contentPadding),
+        horizontalArrangement = Arrangement.Center,
+        verticalAlignment = Alignment.CenterVertically,
+        content = content,
+      )
+    }
+  }
+}
+
+object ButtonDefaults {
```

Comment:
I like the whole system quite a lot! It seems to be quite logical. Maybe something will come up with actual implementation, but now it at least seems very easy-to-use

Reply:
Absolutely, these are for us to figure out how we like them or not. I definitely borrowed this entire structure from material3, where they do more or less the same. I adjusted it to our needs where we gave this 2x2 matrix of possibilities. Size and style. And the public API reflects that exactly. I hope it proves to be as easy to use as I think it will. It will 100% depend also on our designs following the design system properly 😅

Reply:
I think next week you'd have to teach me a bit on how to use it with figma. I guess, for example, surface/opaque/primary is just surfacePrimary in our system, right? But for another example, in figma size: Medium with corner radius/l - this one is not very clear to me yet, how it would be translated to our android.